### PR TITLE
Support add() and diff() for NaiveDateTime

### PR DESCRIPTION
**Pull request by JEG2**
Fixes https://github.com/elixir-lang/elixir/issues/5260

Feedback is more than welcome.  I'm sure I did several silly things here.

### CODE REVIEW

File: lib/elixir/lib/calendar.ex [hunk 1]
```diff
@@ -796,6 +710,75 @@ defmodule NaiveDateTime do
...
+  def add(%NaiveDateTime{} = naive_datetime,
+          integer,
+          unit \\ :seconds) when is_integer(integer) do
+    ndt_microseconds = to_microseconds(naive_datetime)
+    added_microseconds = System.convert_time_unit(integer, unit, :microseconds)
+    sum = ndt_microseconds + added_microseconds
+
+    microsecond = rem(sum, 1_000_000)
+    precision =
+      if microsecond == 0, do: 0, else: Calendar.ISO.precision_for_unit(unit)
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Even if microsecond is zero, it may still have high precision!

For example: `~N[2014-10-02 00:29:10.000000]` is not the same as `~N[2014-10-02 00:29:10]`. One has microseconds precision, the other one does not.

I am also not sure if we should use the `precision_for_unit/1` conversion. For example, if you add microseconds to a naive date time with seconds precision, does it now have microseconds precision? Similarly, if you have a naive date time with microseconds precision and you add seconds to it, does it have seconds precision? I believe the correct approach would be to keep the precision that is already in the NaiveDateTime.

File: lib/elixir/lib/calendar.ex [hunk 2]
```diff
@@ -643,93 +643,7 @@ defmodule NaiveDateTime do
...
-  """
-  @spec from_unix!(non_neg_integer, :native | System.time_unit) :: NaiveDateTime.t
-  def from_unix!(integer, unit \\ :second) when is_atom(unit) do
-    case from_unix(integer, unit) do
-      {:ok, datetime} ->
-        datetime
-      {:error, :invalid_unix_time} ->
-        raise ArgumentError, "invalid Unix time #{integer}"
-    end
+    DateTime.utc_now |> DateTime.to_naive
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Because this code can be used in loops (imagine inserting multiple records in the database), I would like to avoid the double conversion. So what if we still call `Calendar.ISO.from_unix` directly? Something like:

```elixir
{:ok, {year, month, day}, {hour, minute, second}, microsecond} =
  Calendar.ISO.from_unix(:os.system_time, :native)
%NaiveDateTime{year: year, month: month, day: day,
               hour: hour, minute: minute, second: second, microsecond: microsecond}

File: lib/elixir/lib/calendar.ex [hunk 3]
```diff
@@ -796,6 +710,75 @@ defmodule NaiveDateTime do
...
+      # adds seconds by default
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10], 2)
+      ~N[2014-10-02 00:29:12]
+      # accepts negative offsets
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10], -2)
+      ~N[2014-10-02 00:29:08]
+      # can work with other units
+      iex> NaiveDateTime.add(~N[2014-10-02 00:29:10], 2_000, :milliseconds)
+      ~N[2014-10-02 00:29:12]
+      # doesn't lose precision
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: We should likely add tests showing what happens with we use lower precision and higher precision than the one in the NaiveDateTime.

File: lib/elixir/lib/calendar.ex (hunk 1 above)
Comment (JEG2) [inline lib/elixir/lib/calendar.ex]: Just FYI, trying to test this change led to some pretty confusing test output:

```elixir
  1) test doc at NaiveDateTime.add/3 (3) (NaiveDateTimeTest)
     test/elixir/calendar_test.exs:41
     Doctest failed
     code: NaiveDateTime.add(~N[2014-10-02 00:29:10], 21, :milliseconds) === ~N[2014-10-02 00:29:10]                                                                           
           # from gregorian seconds
     left: ~N[2014-10-02 00:29:10]
     stacktrace:
       lib/calendar.ex:725: NaiveDateTime (module)
```

```elixir
  1) test doc at NaiveDateTime.add/3 (3) (NaiveDateTimeTest)
     test/elixir/calendar_test.exs:41
     Doctest failed
     code: NaiveDateTime.add(~N[2014-10-02 00:29:10], 21, :milliseconds) === %NaiveDateTime{calendar: Calendar.ISO, year: 2014, month: 10, day: 2, hour: 0, minute: 29, second: 10, microsecond: {21, 0}}  # ~N[2014-10-02 00:29:10]                                   
           # from gregorian seconds
     left: ~N[2014-10-02 00:29:10]
     stacktrace:
       lib/calendar.ex:725: NaiveDateTime (module)
```

File: lib/elixir/lib/calendar.ex [hunk 4]
```diff
@@ -796,6 +710,75 @@ defmodule NaiveDateTime do
...
+      iex> NaiveDateTime.diff(~N[2014-10-02 00:29:10.042], ~N[2014-10-02 00:29:10.021], :millisecond)
+      21
+      # to gregorian seconds
+      iex> NaiveDateTime.diff(~N[2014-10-02 00:29:10], ~N[0000-01-01 00:00:00])
+      63579428950
+  """
+  @spec diff(NaiveDateTime.t, NaiveDateTime.t, System.time_unit) :: integer
+  def diff(%NaiveDateTime{} = naive_datetime_1,
+           %NaiveDateTime{} = naive_datetime_2,
+           unit \\ :seconds) do
```
Comment (JEG2) [inline lib/elixir/lib/calendar.ex]: Please do double-check me.  I tried to catch them all, but I may have missed something.

Comment (eksperimental) [inline lib/elixir/lib/calendar.ex]: @JEG2 I have. Looks good to me!

File: lib/elixir/lib/calendar/iso.ex [hunk 5]
```diff
@@ -162,9 +162,10 @@ defmodule Calendar.ISO do
     end
   end
 
-  defp precision_for_unit(unit) do
-    subseconds = div System.convert_time_unit(1, :second, unit), 10
-    precision_for_unit(subseconds, 0)
+  @doc false
```
Comment (josevalim) [inline lib/elixir/lib/calendar/iso.ex]: Should we revert this change as we no longer use this function?

**Comment by josevalim** [comment]
Merged! Thank you! :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
